sdk/core/azure-core/src/main/java/com/azure/core/implementation/ByteCountingAsynchronousByteChannel.java

```diff
@@ -0,0 +1,138 @@
+            new DelegatingCompletionHandler<>(future, BYTES_READ_ATOMIC_UPDATER, readProgressReporter));
+        return future;
+    }
+
+    @Override
+    public <A> void write(ByteBuffer src, A attachment, CompletionHandler<Integer, ? super A> handler) {
+        this.channel.write(src, attachment,
+            new DelegatingCompletionHandler<A>(handler, BYTES_WRITTEN_ATOMIC_UPDATER, writeProgressReporter));
+    }
+
+    @Override
+    public Future<Integer> write(ByteBuffer src) {
+        CompletableFuture<Integer> future = new CompletableFuture<>();
+        channel.write(src, src,
+            new DelegatingCompletionHandler<>(future, BYTES_WRITTEN_ATOMIC_UPDATER, writeProgressReporter));
+        return future;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return channel.isOpen();
+    }
+
+    @Override
+    public void close() throws IOException {
+        channel.close();
+    }
+
+    public long getBytesWritten() {
+        return BYTES_WRITTEN_ATOMIC_UPDATER.get(this);
+    }
+
+    public long getBytesRead() {
+        return BYTES_READ_ATOMIC_UPDATER.get(this);
+    }
+
+    private final class DelegatingCompletionHandler<T> implements CompletionHandler<Integer, T> {
+        private final CompletionHandler<Integer, ? super T> handler;
+        private final CompletableFuture<Integer> future;
+        private final AtomicLongFieldUpdater<ByteCountingAsynchronousByteChannel> atomicLongFieldUpdater;
+        private final ProgressReporter progressReporter;
+
+        private DelegatingCompletionHandler(
+            CompletionHandler<Integer, ? super T> handler,
+            AtomicLongFieldUpdater<ByteCountingAsynchronousByteChannel> atomicLongFieldUpdater,
+            ProgressReporter progressReporter) {
+            this.handler = handler;
+            this.future = null;
+            this.atomicLongFieldUpdater = atomicLongFieldUpdater;
+            this.progressReporter = progressReporter;
+        }
+
+        private DelegatingCompletionHandler(
+            CompletableFuture<Integer> future,
+            AtomicLongFieldUpdater<ByteCountingAsynchronousByteChannel> atomicLongFieldUpdater,
+            ProgressReporter progressReporter) {
+            this.handler = null;
+            this.future = future;
+            this.atomicLongFieldUpdater = atomicLongFieldUpdater;
+            this.progressReporter = progressReporter;
+        }
+
+        @Override
+        public void completed(Integer result, T attachment) {
+            if (result > 0) {
+                atomicLongFieldUpdater.addAndGet(ByteCountingAsynchronousByteChannel.this, result);
+                if (progressReporter != null) {
+                    progressReporter.reportProgress(result);
+                }
+            }
+            if (handler != null) {
+                handler.completed(result, attachment);
+            } else if (future != null) {
+                future.complete(result);
+            }
```

Comment:
Does this handle the scenario where the `ByteBuffer` isn't fully written in one write request?

Reply:
It's not a responsibility of this layer. I.e. byte channel contract is that `Integer result` represents how many bytes got written and it's up to caller (for example `IoUtils.transfer` apis) to keep this in mind.

Reply:
see also.

https://github.com/kasobol-msft/azure-sdk-for-java/blob/33a8005500e84dccad9395ca7c253c7885c249ef/sdk/core/azure-core/src/test/java/com/azure/core/implementation/ByteCountingAsynchronousByteChannelTest.java#L129

https://github.com/kasobol-msft/azure-sdk-for-java/blob/33a8005500e84dccad9395ca7c253c7885c249ef/sdk/core/azure-core/src/main/java/com/azure/core/util/IOUtils.java#L106-L110

Reply:
> It's not a responsibility of this layer. I.e. byte channel contract is that `Integer result` represents how many bytes got written and it's up to caller (for example `IoUtils.transfer` apis) to keep this in mind.

Got it, was wondering that, might want to document that for clarification.